Add OnCompletion Hook to ProcessGroup

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106988
* #107160
* #107141
* #107140

This allows infra/trainers to get detailed stats about communication
efficiencies without know anything about what model or distributed
training paradigms have been used. This is helpful as infra/trainer
package usually prefers to be as model/algorithm agnostic as possible.
Therefore, we cannot assume that infra/trainer can have access to all
collectives used by the model authors.

This commit adds an `OnCompletion` hook to `ProcessGroupNCCL` which
will be fired on every work completion event.

mrshenli committed:
Add OnCompletion Hook to ProcessGroup

[ghstack-poisoned]

rohan-varma reviewed (commented):
at a high level, curious how this is different from registering a callback on `work.get_future()`?

mrshenli commented:
Hey @rohan-varma, this one is for infra/trainer devs, where they usually treat model/parallelism as a black box, and therefore do not have access to the `work` object and might not even have any idea on which collectives are used under the hood. However, they still need to know the efficiency of the job in general. Hence, this PR adds that directly to the PG. 

For context, we are seeing some workloads suffer from straggler problems. Before diving deeper into models and develop solutions, we would like to know how much speedup we can get from resolving stragglers. And therefore, would like to know some more lower-level information. :)

mrshenli commented:
Let me clean up the PR and add some more details to the summary.

mrshenli committed:
Update on "Add OnCompletion Hook to ProcessGroup"


This allows infra/trainers to get detailed stats about communication
efficiencies without know anything about what model or distributed
training paradigms have been used. This is helpful as infra/trainer
package usually prefers to be as model/algorithm agnostic as possible.
Therefore, we cannot assume that infra/trainer can have access to all
collectives used by the model authors.

This commit adds an `OnCompletion` hook to `ProcessGroupNCCL` which
will be fired on every work completion event.

[ghstack-poisoned]

mrshenli committed:
Update on "Add OnCompletion Hook to ProcessGroup"


This allows infra/trainers to get detailed stats about communication
efficiencies without know anything about what model or distributed
training paradigms have been used. This is helpful as infra/trainer
package usually prefers to be as model/algorithm agnostic as possible.
Therefore, we cannot assume that infra/trainer can have access to all
collectives used by the model authors.

This commit adds an `OnCompletion` hook to `ProcessGroupNCCL` which
will be fired on every work completion event.

[ghstack-poisoned]

mrshenli committed:
Update on "Add OnCompletion Hook to ProcessGroup"


This allows infra/trainers to get detailed stats about communication
efficiencies without know anything about what model or distributed
training paradigms have been used. This is helpful as infra/trainer
package usually prefers to be as model/algorithm agnostic as possible.
Therefore, we cannot assume that infra/trainer can have access to all
collectives used by the model authors.

This commit adds an `OnCompletion` hook to `ProcessGroupNCCL` which
will be fired on every work completion event.

[ghstack-poisoned]

H-Huang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -886,68 +929,122 @@ void ProcessGroupNCCL::logWorkEnd(WorkNCCL& work) {
+        completedWorkList.back().handleException(asyncErrorHandling_);
       }
+      // Abort all NCCL Communicators on this ProcessGroupNCCL instance.
+      abort();
```
in what case would `completedWorkList` be empty / do we ever get to this `abort()`?

mrshenli replied:
It should never get here, but feels un-secure if we just ignore the else branch. Let me add an `abortReason` argument to clarify in case somehow this branch is hit.

mrshenli replied:
addressed

H-Huang reviewed (commented):
Overall looks good to me, just some small questions and comments. I'm okay with merging this in after comments discussion and thanks for adding Shen!

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2454,6 +2454,199 @@ def test_channels_last_contig(self):
+        num_hook_fired = 0
+        durations: List[float] = []
+
+        def hook(work):
```
does work obj contain metadata about the op, the input sizes, the PG/other ranks?  i'm guessing yes but just to be sure, seems like all of that would be useful to have access to from the hook

mrshenli replied:
Yep, `WorkNCCL` has all of those except an direct access to PG, but we can add that if necessary either as a reference in Work or as an argument to the hook.

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2454,6 +2454,199 @@ def test_channels_last_contig(self):
+            num_hook_fired += 1
+            durations.append(work._get_duration())
+
+        pg._register_on_completion_hook(hook)
```
building on your PR desc saying that trainer may not have insight into model / collectives used, should we assume the same thing about PG creation?  I wonder if its valuable to have a higher-level hook installed globally on c10d lib by the trainer that would log work events across PGs.

(if you agree, i think this could still be a separate PR and probably build on top of the infra in this PR)

mrshenli replied:
> building on your PR desc saying that trainer may not have insight into model / collectives used, should we assume the same thing about PG creation? 

Good point! It's less intrusive to get PGs compared to every Work, but it's certainly better if infra/trainers do not have to. 

> I wonder if its valuable to have a higher-level hook installed globally on c10d lib by the trainer that would log work events across PGs.

Sure, I can add that in a followup PR.

kumpera commented on torch/csrc/distributed/c10d/ProcessGroup.hpp:
```diff
@@ -663,6 +663,16 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
 
+
+  void registerOnCompletionHook(std::function<void(c10::intrusive_ptr<Work>)>&& hook) {
+    getDefaultBackend()->registerOnCompletionHook(std::move(hook));
```
Shouldn't we register with all backends?

kumpera commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -783,6 +796,36 @@ uint64_t ProcessGroupNCCL::getSequenceNumberForGroup() {
+      "ProcessGroupNCCL OnCompletion hook requires recording start and end "
+      "events which require setting NCCL_ENABLE_TIMING environment variable. "
+      "This is only available for NCCL version >= 2.4.");
+  onCompletionHook_ = std::move(hook);
```
This is a fragile design, can't enable recording at this point?

mrshenli replied:
ah, yep, good point! Will keep just one hook in this PR and add a check to prevent re-registering. Not sure if we should allow multiple hooks in the future. We can revisit if necesssary.

kumpera replied:
I believe that having an env var gate an optional feature is fragile. It will lead to a lot of extra work from our users when they get the exception.

Why can't we enable recording those events when the callback is set and not have this env var at all?

mrshenli replied:
> Why can't we enable recording those events when the callback is set and not have this env var at all?

The original reason what that users might not need enableTiming if they don't need CUDA elapsed time. But I guess that can also be done with a separate argument when register the hook.

Let me make the change to avoid the env var.

mrshenli replied:
> Why can't we enable recording those events when the callback is set and not have this env var at all?

Hey @kumpera,  2nd thought on this one. As users might register the hook in the middle of a job where there could be pending collective work items created with `enableTiming=False` and then toggled to true. I.e., users can get duration on some work but not others, which is confusing. This means we need a flush when register hooks and block future collectives until registration is done. I'd prefer to defer this design decision to future PRs. 

I am also hesitate to turn on recording and timing for all start and end events by default, as changing this default behavior has a large blast radius. 

Another option is to add hooks to the pg options arg for PG initialization. However, this API does not align where with other parts of PyTorch. 

thoughts? :)

kumpera replied:
I agree that recording by default requires being a lot more careful.

Registering hooks in the middle of a job will be confusing regardless. For example, do we fire the hook only for later collectives or is it ok to racy'ish fire for currently running ones?

mrshenli replied:
> Registering hooks in the middle of a job will be confusing regardless. For example, do we fire the hook only for later collectives or is it ok to racy'ish fire for currently running ones?

Not sure yet. I'd vote for we should fire hooks only for later collectives. Or, maybe we should ban this, where hooks can only be registered before firing any collectives, otherwise error out?

kumpera commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -886,68 +929,122 @@ void ProcessGroupNCCL::logWorkEnd(WorkNCCL& work) {
+        // with an intrusive_ptr so that pybind can safely pass by copy
+        // and own the ptr lifetime.
+        onCompletionHook_(
+            c10::make_intrusive<WorkNCCL>(completedWorkList.back()));
```
Why are we copying the Work object?

mrshenli replied:
`Work` does not have move ctor. Let me know if you want me to add that.

kumpera reviewed (changes requested):
I think it's a bad idea to directly call user callback from the watchdog threads.
Since the work object we produce is non functional, we should simply give the user a map with (start,end,sequence,collective) keys.

mrshenli commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2454,6 +2454,199 @@ def test_channels_last_contig(self):
+        self._spawn_processes()
+
+    def tearDown(self):
+        super().tearDown()
```
oh, yep, good catch!

mrshenli replied:
addressed

mrshenli commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -314,20 +314,28 @@ ProcessGroupNCCL::WorkNCCL::WorkNCCL(
+    ncclStartEvents_ = std::make_shared<std::vector<at::cuda::CUDAEvent>>();
+    ncclStartEvents_->reserve(devices.size());
+    for (uint32_t i = 0; i < devices.size(); ++i) {
+      ncclStartEvents_->emplace_back(at::cuda::CUDAEvent(cudaEventDefault));
```
for `desyncDebug` mode, these events are already there, but this PR turns on enableTiming for those events to get the exact duration of a collective. My intention was to have an accurate measure of the collective operator, i.e., launched-to-finished instead of enqueued-to-finished. Let me know if there are better options to achieve this. :)

mrshenli commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -632,6 +641,9 @@ ProcessGroupNCCL::ProcessGroupNCCL(
       parseEnvVarIntDefault(NCCL_ASYNC_ERROR_HANDLING, 1 /*TearDown*/));
   desyncDebug_ = parseEnvVarFlag(NCCL_DESYNC_DEBUG) ||
       (dist_debug_level_ >= DebugLevel::Detail);
+#ifdef ENABLE_NCCL_ERROR_CHECKING
```
Makes sense!

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1380,6 +1380,10 @@ def destroy_process_group(group: Optional[ProcessGroup] = None):
 
+    # prevent Python Interpreter to exit before all hooks are fired
+    if pg.name().lower() == "nccl":
+        pg._wait_for_pending_works()
```
It's likely that Python interpreter will exit before Cpp destructs PG instance and joins the watchdog thread. As a result, the Python hooks fired in watchdog thread will crash when attempting to grab GIL. So we basically have two options:

1. revive Python Interpreter if necessary before running hook
2. keep the interpreter alive until all hooks are fired. 

This PR attempts to follow options 2. Let me know if we prefer a different solution. 

Let me add more comments in code to explain this.

kumpera replied:
A safer design would be to have all watchdog threads enqueue events (and not work objects) to a MPSC queue and then have a python thread dequeueing from them.

This would further improve isolation from callback issues.

mrshenli replied:
Hey @kumpera,  you mean start a thread in Python and call into C++ to release GIL? 

The current version spins up a dedicated thread from C++ to handle hooks (in parallel to NCCL watchDog). This PR is getting a bit too long. I'd prefer introduce more features in followup PRs if that's OK to you.

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -32,6 +32,7 @@
 
 #include <fmt/format.h>
 #include <pybind11/chrono.h>
+#include <pybind11/embed.h>
```
debugging vestige, will delete

mrshenli commented on torch/csrc/distributed/c10d/Work.cpp:
```diff
@@ -127,6 +127,11 @@ void Work::finishAndThrow(std::exception_ptr exception) {
 
+float Work::getDuration() const {
+  TORCH_CHECK(
+      false, "Only ProcessGrouppNCCL::WorkNCCL supports getDuration, but got ");
```
will fix message

mrshenli reviewed (commented):
Hey @rodrigobdz, thanks for the review!

> I think it's a bad idea to directly call user callback from the watchdog threads.

Do you prefer a dedicated thread for it instead?

> Since the work object we produce is non functional, we should simply give the user a map with (start,end,sequence,collective) keys.

Sure, I can create a separate struct just for the hook.

mrshenli committed:
Update on "Add OnCompletion Hook to ProcessGroup"


This allows infra/trainers to get detailed stats about communication
efficiencies without know anything about what model or distributed
training paradigms have been used. This is helpful as infra/trainer
package usually prefers to be as model/algorithm agnostic as possible.
Therefore, we cannot assume that infra/trainer can have access to all
collectives used by the model authors.

This commit adds an `OnCompletion` hook to `ProcessGroupNCCL` which
will be fired on every work completion event.

[ghstack-poisoned]

mrshenli committed:
Update on "Add OnCompletion Hook to ProcessGroup"


This allows infra/trainers to get detailed stats about communication
efficiencies without know anything about what model or distributed
training paradigms have been used. This is helpful as infra/trainer
package usually prefers to be as model/algorithm agnostic as possible.
Therefore, we cannot assume that infra/trainer can have access to all
collectives used by the model authors.

This commit adds an `OnCompletion` hook to `ProcessGroupNCCL` which
will be fired on every work completion event.

[ghstack-poisoned]

mrshenli reviewed (commented):
Hey @H-Huang, @wconstab, @kumpera, thanks again for reviewing. Could you please help take another look. Main changes since last time are:

1. Instead of passing `Work` directly to user hook, we now use a read-only `WorkInfo` struct to pass op type, start time (when user calls collective), finish time (detected by watch dog thread), and active duration (from CUDA elapsed time). 
2. Instead of calling hook within the watch dog thread, we now use a dedicated c++ thread to run hooks. There is on-going discussion whether we should use a Python thread instead and rely on a Multi-Producer Single-Consumer queue to run those hooks. I'd prefer to leave it to future PRs to figure out what's the optimal way to run user hooks. This PR aims to unblock a use case quickly without altering default PG NCCL behavior when there is no hooks.

mrshenli committed:
Update on "Add OnCompletion Hook to ProcessGroup"


This allows infra/trainers to get detailed stats about communication
efficiencies without know anything about what model or distributed
training paradigms have been used. This is helpful as infra/trainer
package usually prefers to be as model/algorithm agnostic as possible.
Therefore, we cannot assume that infra/trainer can have access to all
collectives used by the model authors.

This commit adds an `OnCompletion` hook to `ProcessGroupNCCL` which
will be fired on every work completion event.

[ghstack-poisoned]

kumpera commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -886,71 +938,152 @@ void ProcessGroupNCCL::logWorkEnd(WorkNCCL& work) {
+
+    {
+      const std::lock_guard<std::mutex> lock(completedWorkListMutex_);
+      completedWorkList_.splice(completedWorkList_.end(), completedWorkList);
```
you need to notify `completedWorkListCV_`.

kumpera reviewed (approved):
LGTM.

This is a great starting point. The only issue I see is that the callback thread CV doesn't get notified.

H-Huang reviewed (approved):
LGTM, thanks for updating :)

mrshenli committed:
Update on "Add OnCompletion Hook to ProcessGroup"


This allows infra/trainers to get detailed stats about communication
efficiencies without know anything about what model or distributed
training paradigms have been used. This is helpful as infra/trainer
package usually prefers to be as model/algorithm agnostic as possible.
Therefore, we cannot assume that infra/trainer can have access to all
collectives used by the model authors.

This commit adds an `OnCompletion` hook to `ProcessGroupNCCL` which
will be fired on every work completion event.

[ghstack-poisoned]

mrshenli committed:
Update on "Add OnCompletion Hook to ProcessGroup"


This allows infra/trainers to get detailed stats about communication
efficiencies without know anything about what model or distributed
training paradigms have been used. This is helpful as infra/trainer
package usually prefers to be as model/algorithm agnostic as possible.
Therefore, we cannot assume that infra/trainer can have access to all
collectives used by the model authors.

This commit adds an `OnCompletion` hook to `ProcessGroupNCCL` which
will be fired on every work completion event.

[ghstack-poisoned]

mrshenli committed:
Update on "Add OnCompletion Hook to ProcessGroup"


This allows infra/trainers to get detailed stats about communication
efficiencies without know anything about what model or distributed
training paradigms have been used. This is helpful as infra/trainer
package usually prefers to be as model/algorithm agnostic as possible.
Therefore, we cannot assume that infra/trainer can have access to all
collectives used by the model authors.

This commit adds an `OnCompletion` hook to `ProcessGroupNCCL` which
will be fired on every work completion event.

[ghstack-poisoned]

mrshenli commented:
@pytorchbot merge -f "windows and rocm test failures are irrelevant"

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting you change, but it is failing Windows build with some linker error.  The Windows failures on PR looks legit' -c ignoredsignal

```
2023-08-15T05:57:10.4987572Z FAILED: bin/torch_python.dll lib/torch_python.lib 
2023-08-15T05:57:10.4989481Z cmd.exe /C "cd . && C:\Jenkins\Miniconda3\Library\bin\cmake.exe -E vs_link_dll --intdir=caffe2\torch\CMakeFiles\torch_python.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\BUILDT~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\torch_python.rsp  /out:bin\torch_python.dll /implib:lib\torch_python.lib /pdb:bin\torch_python.pdb /dll /version:0.0 /machine:x64 /ignore:4049 /ignore:4217 /ignore:4099 /INCREMENTAL:NO  /NODEFAULTLIB:LIBCMT.LIB  -WHOLEARCHIVE:C:/actions-runner/_work/pytorch/pytorch/build/lib/onnx.lib  && cd ."
2023-08-15T05:57:10.4992301Z LINK: command "C:\PROGRA~2\MICROS~1\2019\BUILDT~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\torch_python.rsp /out:bin\torch_python.dll /implib:lib\torch_python.lib /pdb:bin\torch_python.pdb /dll /version:0.0 /machine:x64 /ignore:4049 /ignore:4217 /ignore:4099 /INCREMENTAL:NO /NODEFAULTLIB:LIBCMT.LIB -WHOLEARCHIVE:C:/actions-runner/_work/pytorch/pytorch/build/lib/onnx.lib /MANIFEST /MANIFESTFILE:bin\torch_python.dll.manifest" failed (exit code 1120) with the following output:
2023-08-15T05:57:10.4993784Z    Creating library lib\torch_python.lib and object lib\torch_python.exp
2023-08-15T05:57:10.4996429Z init.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl c10d::PythonOnCompletionHook::PythonOnCompletionHook(class pybind11::object)" (__imp_??0PythonOnCompletionHook@c10d@@QEAA@Vobject@pybind11@@@Z) referenced in function "public: struct pybind11::detail::void_type __cdecl pybind11::detail::argument_loader<class c10::intrusive_ptr<class c10d::ProcessGroup,struct c10::detail::intrusive_target_default_null_type<class c10d::ProcessGroup> > const &,class pybind11::object>::call<void,class pybind11::gil_scoped_acquire,class <lambda_c8ac4a37dc06d3431128ad7dfdd94e65> &>(class <lambda_c8ac4a37dc06d3431128ad7dfdd94e65> &)&& " (??$call@XVgil_scoped_acquire@pybind11@@AEAV<lambda_c8ac4a37dc06d3431128ad7dfdd94e65>@@@?$argument_loader@AEBV?$intrusive_ptr@VProcessGroup@c10d@@U?$intrusive_target_default_null_type@VProcessGroup@c10d@@@detail@c10@@@c10@@Vobject@pybind11@@@detail@pybind11@@QEHAA?AUvoid_type@12@AEAV<lambda_c8ac4a37dc06d3431128ad7dfdd94e65>@@@Z)
2023-08-15T05:57:10.5064986Z init.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl c10d::PythonOnCompletionHook::~PythonOnCompletionHook(void)" (__imp_??1PythonOnCompletionHook@c10d@@QEAA@XZ) referenced in function "public: __cdecl std::function<void __cdecl(class std::shared_ptr<struct c10d::WorkInfo>)>::function<void __cdecl(class std::shared_ptr<struct c10d::WorkInfo>)><class <lambda_bc550a0d221355b3e81c21183caeeaef>,0>(class <lambda_bc550a0d221355b3e81c21183caeeaef>)" (??$?0V<lambda_bc550a0d221355b3e81c21183caeeaef>@@$0A@@?$function@$$A6AXV?$shared_ptr@UWorkInfo@c10d@@@std@@@Z@std@@QEAA@V<lambda_bc550a0d221355b3e81c21183caeeaef>@@@Z)
2023-08-15T05:57:10.5067821Z init.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: void __cdecl c10d::PythonOnCompletionHook::operator()(class std::shared_ptr<struct c10d::WorkInfo>)const " (__imp_??RPythonOnCompletionHook@c10d@@QEBAXV?$shared_ptr@UWorkInfo@c10d@@@std@@@Z) referenced in function "public: static void __cdecl std::_Invoker_ret<void,1>::_Call<class <lambda_bc550a0d221355b3e81c21183caeeaef> &,class std::shared_ptr<struct c10d::WorkInfo> >(class <lambda_bc550a0d221355b3e81c21183caeeaef> &,class std::shared_ptr<struct c10d::WorkInfo> &&)" (??$_Call@AEAV<lambda_bc550a0d221355b3e81c21183caeeaef>@@V?$shared_ptr@UWorkInfo@c10d@@@std@@@?$_Invoker_ret@X$00@std@@SAXAEAV<lambda_bc550a0d221355b3e81c21183caeeaef>@@$$QEAV?$shared_ptr@UWorkInfo@c10d@@@1@@Z)
2023-08-15T05:57:10.5070579Z init.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl c10d::PythonOnCompletionHook::PythonOnCompletionHook(class c10d::PythonOnCompletionHook const &)" (__imp_??0PythonOnCompletionHook@c10d@@QEAA@AEBV01@@Z) referenced in function "public: __cdecl std::_Func_impl_no_alloc<class <lambda_bc550a0d221355b3e81c21183caeeaef>,void,class std::shared_ptr<struct c10d::WorkInfo> >::_Func_impl_no_alloc<class <lambda_bc550a0d221355b3e81c21183caeeaef>,void,class std::shared_ptr<struct c10d::WorkInfo> ><class <lambda_bc550a0d221355b3e81c21183caeeaef> const &,0>(class <lambda_bc550a0d221355b3e81c21183caeeaef> const &)" (??$?0AEBV<lambda_bc550a0d221355b3e81c21183caeeaef>@@$0A@@?$_Func_impl_no_alloc@V<lambda_bc550a0d221355b3e81c21183caeeaef>@@XV?$shared_ptr@UWorkInfo@c10d@@@std@@@std@@QEAA@AEBV<lambda_bc550a0d221355b3e81c21183caeeaef>@@@Z)
2023-08-15T05:57:10.5072460Z init.cpp.obj : error LNK2001: unresolved external symbol "public: __cdecl c10d::PythonOnCompletionHook::~PythonOnCompletionHook(void)" (??1PythonOnCompletionHook@c10d@@QEAA@XZ)
2023-08-15T05:57:10.5073155Z bin\torch_python.dll : fatal error LNK1120: 5 unresolved externals
```

huydhn commented:
Here is the failure in trunk https://hud.pytorch.org/pytorch/pytorch/commit/ba1da47e8fa95ca0dd8b2d63430f7eb54fdbbccb
